airflow/jobs.py

```diff
@@ -598,171 +866,475 @@ def prioritize_queued(self, session, executor, dagbag):
+        5. Kill (in ORM) any task instances belonging to the DAGs that haven't
+        issued a heartbeat in a while.
+
+        Returns a list of SimpleDag objects that represent the DAGs found in
+        the file
+
+        :param file_path: the path to the Python file that should be executed
+        :type file_path: unicode
+        :param pickle_dags: whether serialize the DAGs found in the file and
+        save them to the db
+        :type pickle_dags: bool
+        :return: a list of SimpleDags made from the Dags found in the file
+        :rtype: list[SimpleDag]
+        """
+        self.logger.info("Processing file {} for tasks to queue".format(file_path))
+        # As DAGs are parsed from this file, they will be converted into SimpleDags
+        simple_dags = []
+
+        TI = models.TaskInstance
+
+        try:
+            dagbag = models.DagBag(file_path)
+        except Exception as e:
+            self.logger.exception("Failed at reloading the DAG file {}".format(file_path))
+            Stats.incr('dag_file_refresh_error', 1, 1)
+            return []
+
+        if len(dagbag.dags) > 0:
+            self.logger.info("DAG(s) {} retrieved from {}"
+                             .format(dagbag.dags.keys(),
+                                     file_path))
+        else:
+            self.logger.warn("No viable dags retrieved from {}".format(file_path))
+            return []
+
+        # Save individual DAGs in the ORM
+        sync_time = datetime.now()
+        for dag in dagbag.dags.values():
+            models.DAG.sync_to_db(dag, sync_time)
+
+        # Pickle the DAGs (if necessary) and put them into a SimpleDag
+        for dag_id in dagbag.dags:
+            dag = dagbag.get_dag(dag_id)
+            pickle_id = None
+            if pickle_dags:
+                pickle_id = dag.pickle(session).id
+
+            task_ids = [task.task_id for task in dag.tasks]
+            simple_dags.append(SimpleDag(dag.dag_id,
+                                         task_ids,
+                                         dag.full_filepath,
+                                         dag.concurrency,
+                                         dag.is_paused,
+                                         pickle_id))
+
+        if len(self.dag_ids) > 0:
+            dags = [dag for dag in dagbag.dags.values() if dag.dag_id in self.dag_ids]
+        else:
+            dags = [
+                dag for dag in dagbag.dags.values()
+                if not dag.parent_dag]
+
+        tis_q = multiprocessing.Queue()
+
+        self._process_dags(dagbag, dags, tis_q)
+
+        while not tis_q.empty():
+            ti_key = tis_q.get()
+            dag = dagbag.dags[ti_key[0]]
+            task = dag.get_task(ti_key[1])
+            ti = TI(task, ti_key[2])
+            # Task starts out in the queued state. All tasks in the queued
+            # state will be scheduled later in the execution loop.
+            ti.state = State.QUEUED
```

Comment:
(A quick overview of task states. This isn't necessarily how things *should be* but it is how they *are* :) )

Indeed -- all tasks used to start with `State.NONE`. @bolkedebruin's latest PR introduces `State.SCHEDULED` when the TI has been created by the scheduler but not yet run. When the executor loads a TI, it eventually calls TI.run() which does one of two things:
1. If the TI can be run, sets `State.RUNNING` and calls the `execute()` method.
2. If the TI has a pool, sets `State.QUEUED` and returns
The scheduler has a `prioritize_queued` method which loads up all the queued tasks and tries to run them if there are slots available in their respective pools. That second run is the one that actually moves pooled tasks to `State.RUNNING`.